pkg/controller/controller.go

```diff
@@ -52,6 +52,8 @@ type AddArgs struct {
 	ControllerOptions controller.Options
 	// Type is the object type to watch.
 	Type client.Object
+	// ShouldEnsureDeleted specifies that the controller should ensure the object is properly deleted on not found.
```

Comment:
Why do we need that configurable? Should the controller not always ensure to delete object in case it is not found?

Reply:
I only wanted to have this for the service and node controllers, to ensure that their corresponding "owned" objects are properly deleted no matter what. This is more a safety precaution rather than an essential mechanism, e.g. if a virtualmachine is there but its node is gone (which should in theory never happen but in practice somehow does happen, e.g. due to manual intervention), upon restart the non-existing node will be reconciled and the virtualmachine will be deleted as a result. For the publicipaddress and virtualmachine controllers this is not needed and would only result in unneeded reconcilations and "noise". Besides, it does impose additional requirements to `Delete` that have to be fulfilled, see the other comments below.